Refactor SILDeclRef constructors

JaSpa opened the pull request:
This addresses a FIXME and reduces code/logic duplication in the `SILDeclRef` constructors.

JaSpa committed:
Refactor multiple getNaturalUncurryLevel implementations into one

There were two different places which did this
calculation, let them call the same function.

gottesmm commented:
Meta comment:

1. ec7736b seems to be doing a couple of things at once. Each one of those things should be different commits and also need more descriptive commit messages.
2. We generally do not use the term "DRY" (I have never seen people use it TBH). If you are refactoring out common code, I would just say that. That makes the message more precise.

gottesmm commented on include/swift/AST/AnyFunctionRef.h:
```diff
@@ -51,6 +51,15 @@ class AnyFunctionRef {
     }
   }
 
+  // Get the natural uncurry level for the embedded function which is the number
```
This should use 3 slashes. See the LLVM style guide:

http://llvm.org/docs/CodingStandards.html

gottesmm commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -438,16 +438,9 @@ emitRValueForDecl(SILLocation loc, ConcreteDeclRef declRef, Type ncRefType,
-      ++uncurryLevel;
-  }
-
+  if (auto fd = dyn_cast<FuncDecl>(decl))
```
Can you put back in the * here?

http://llvm.org/docs/CodingStandards.html#beware-unnecessary-copies-with-auto

JaSpa committed:
Function for validating the uncurry level & setting the isCurried flag

JaSpa committed:
Assign fields from constructor arguments at the top

This saves us from assigning the same value to the
same field on different paths.

JaSpa committed:
Factor the big if-statements in the constructors out

Their logic is now unified in one function which, based on the `setKind`
flag, either sets this reference to a matching `SILDeclRef::Kind` or
checks that the given kind is valid for the given `ValueDecl`.

JaSpa commented:
I've split the big commit into multiple smaller, I hope it's more clear now what I'm doing.
… and I've revised my commit messages, I don't know where I've heard “DRY out“ before but it's from the [Don't repeat yourself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle.

JaSpa committed:
Put `*` after auto to guard against implicit copies.

gottesmm commented on lib/SIL/SILDeclRef.cpp:
```diff
@@ -112,129 +112,100 @@ bool swift::requiresForeignEntryPoint(ValueDecl *vd) {
-    : atUncurryLevel;
-  isCurried = uncurryLevel != naturalUncurryLevel;
+  default:
+    llvm_unreachable("unhandled kind of ValueDecl for SILDeclRef");
```
The point of changing this to be a switch is to create a covered switch so we get a warning when we add a new DeclKind.

The way to do this correctly is to eliminate the default case and move the unreachable outside the switch.

JaSpa replied:
Ok, I think I understand now, what you're saying.
But for a covered switch I have to list all other `DeclKind`s, right? Which means Either `break` from the switch and then `llvm_unreachable` or `llvm_unreachable` instead of `break`.

gottesmm reviewed (changes requested):
Small change

gottesmm commented:
Fix that and then this looks reasonable to me. I dislike some of the using of closures, but from what I understand at some point uncurry levels will be removed. This at least makes it easier to catch changes that add new DeclKind.

JaSpa committed:
Turn the switch into a covered switch

This produces a warning when introducing a new
ValueDecl without updating this switch.

gottesmm commented on lib/SIL/SILDeclRef.cpp:
```diff
@@ -162,8 +162,22 @@ static unsigned getNaturalUncurryLevelAndKind(ValueDecl *vd,
+#define VALUE_DECL(...)
+#define DECL(Id, Parent) case DeclKind::Id:
+#include "swift/AST/DeclNodes.def"
+    llvm_unreachable("impossible value decl");
```
Question. How many of these are you including? By using the include file, you are not going to get the covered switch warning. I think you really need to handle /all/ cases directly. The whole point is if someone adds something to the DeclNodes.def file, we want them to use the covered switch warning to think about what needs to be done here.

JaSpa commented:
The macros in the include file used for subclasses of `ValueDecl` all delegate back to the `VALUE_DECL` macro for which I actively disable the replacement using the empty define.
As the argument to the function is `ValueDecl`, we should never get non-`ValueDecl`s, which will never require manual handling.
In other words, if a new non-`ValueDecl` is added, this won't matter for this code here, because it should never get there. If a new `ValueDecl` subclass is added, we *will* get a warning, because it's not explicitly handled, not even by the include because of the empty define.
I hope this clarifies my approach.

Von meinem iPhone gesendet

> Am 27.10.2016 um 18:45 schrieb Michael Gottesman <notifications@github.com>:
> 
> @gottesmm requested changes on this pull request.
> 
> In lib/SIL/SILDeclRef.cpp:
> 
> > +
> +  case DeclKind::GenericTypeParam:
> +  case DeclKind::AssociatedType:
> +  case DeclKind::TypeAlias:
> +  case DeclKind::Enum:
> +  case DeclKind::Struct:
> +  case DeclKind::Protocol:
> +  case DeclKind::Module:
> +  case DeclKind::Param:
> +  case DeclKind::Subscript:
> +    llvm_unreachable("invalid kind of ValueDecl for SILDeclRef");
> +
> +#define VALUE_DECL(...)
> +#define DECL(Id, Parent) case DeclKind::Id:
> +#include "swift/AST/DeclNodes.def"
> +    llvm_unreachable("impossible value decl");
> Question. How many of these are you including? By using the include file, you are not going to get the covered switch warning. I think you really need to handle /all/ cases directly. The whole point is if someone adds something to the DeclNodes.def file, we want them to use the covered switch warning to think about what needs to be done here.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slavapestov commented:
Hi @JaSpa, I recently removed uncurryLevel from SILDeclRef -- does the rest of this PR still matter? Sorry we dropped the ball on this months ago.

CodaFi commented:
This one hasn't been touched in quite a while and has fallen behind some refactorings in this area.  Going to close this.  Feel free to reopen it and rebase it if that is not the case.
